File: backend/app/models/maintenance.py
from sqlalchemy import (
    Column,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base
from app.models.enums import MaintenanceStatus


class MaintenanceLog(Base):
    __tablename__ = "maintenance_logs"

    id = Column(Integer, primary_key=True, index=True)

    vehicle_id = Column(
        Integer,
        ForeignKey("vehicles.id"),
        nullable=False
    )

    issue = Column(
        String(200),
        nullable=False
    )

    description = Column(Text)

    maintenance_cost = Column(
        Numeric(12,2),
        default=0
    )

    start_date = Column(Date)

    end_date = Column(Date)

    status = Column(
        Enum(MaintenanceStatus),
        default=MaintenanceStatus.PENDING,
        nullable=False
    )

    created_by = Column(
        Integer,
        ForeignKey("users.id")
    )

    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now()
    )

    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now()
    )

    vehicle = relationship(
        "Vehicle",
        back_populates="maintenance_logs"
    )

    def __repr__(self):
        return f"<Maintenance {self.id}>"

File: frontend/src/services/maintenance.js
import api from "./api";

export function normalizeMaintenance(record) {
  return {
    ...record,
    vehicle_id: Number(record.vehicle_id),
    maintenance_cost: Number(record.maintenance_cost || 0),
  };
}

export function toMaintenancePayload(values) {
  return {
    vehicle_id: Number(values.vehicle_id),
    issue: values.issue,
    description: values.description || null,
    maintenance_cost: String(values.maintenance_cost || 0),
    start_date: values.start_date || null,
    end_date: values.end_date || null,
    status: values.status || "Pending",
    created_by: values.created_by || null,
  };
}

export async function listMaintenance(params = {}) {
  const { data } = await api.get("/maintenance/", { params });
  return data.map(normalizeMaintenance);
}

export async function createMaintenance(values) {
  const { data } = await api.post("/maintenance/", toMaintenancePayload(values));
  return normalizeMaintenance(data);
}

export async function updateMaintenance(id, values) {
  const { data } = await api.put(`/maintenance/${id}`, toMaintenancePayload(values));
  return normalizeMaintenance(data);
}

export async function deleteMaintenance(id) {
  await api.delete(`/maintenance/${id}`);
}
